### PR TITLE
Adding a CONTRIBUTING page

### LINKED ISSUE #1781: README.md is missing instructions for contributors
The README just mentions how to install Dafny from pre-built binaries, but does not point to building from source or how to contribute (https://github.com/dafny-lang/dafny/wiki/INSTALL#building-and-developing-from-source-code). The only instructions are for how to install `pre-commit` which seem out of place. Maybe there's place for a CONTRIBUTING.md.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #1781

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+## Contributing to Dafny
+
+Dafny is an Open Source project and welcomes contributions.
+
+Developers who would like to experiment with and perhaps modify Dafny source should note these resources:
+- First, there are [resources to learn about Dafny as a user](https://dafny.org/dafny/).
+- And note the [community Code of Conduct](https://github.com/dafny-lang/dafny/blob/master/.github/CODE_OF_CONDUCT.md) expected of all participating in the project.
+- The Dafny project is at [github](https://github.com/dafny/dafny-lang) with related [sibling repositories](https://github.com/dafny).
```
Comment (robin-aws) [inline CONTRIBUTING.md:8]:
```suggestion
- The Dafny project is at [github](https://github.com/dafny/dafny-lang) with related [sibling repositories](https://github.com/dafny-lang).
```

File: README.md [hunk 2]
```diff
@@ -70,6 +70,8 @@ The language itself draws pieces of influence from:
 
 # Contributors
 
+Information and instructions for potential contributors are provided [here](CONTRIBUTING).
+
```
Comment (robin-aws) [inline README.md:74]: The information on the pre-commit hook should be moved to the existing `.github/CONTRIBUTING.md` file as well.

File: README.md [hunk 3]
```diff
@@ -70,6 +70,8 @@ The language itself draws pieces of influence from:
 
 # Contributors
 
+Information and instructions for potential contributors are provided [here](CONTRIBUTING).
```
Comment (robin-aws) [inline README.md:73]:
```suggestion
Information and instructions for potential contributors are provided [here](.github/CONTRIBUTING.md).
```

**Review by robin-aws — CHANGES_REQUESTED** [review]
I got halfway through reviewing this before realizing/recalling that we already have a `.github/CONTRIBUTING.md` file!

The GitHub documentation is not opinionated about which path is better, so we might as well stick with the existing one. Feel free to add any extra content you have here in that file instead, and it's definitely a good change to add a link to the main README file so other readers don't miss it like we both did. :)

**Comment by MikaelMayer** [comment]
> I got halfway through reviewing this before realizing/recalling that we already have a `.github/CONTRIBUTING.md` file!

I got confused at first too when I created a similar PR. I really think CONTRIBUTING.md should not be part of a vendor `.github` folder but at the top level folder too.

**Comment by robin-aws** [comment]
Having slept on it and observed two more Dafny contributors miss the existing file, I say we move it to the root folder :)

**Comment by davidcok** [comment]
My thought also -- and will do

File: README.md (hunk 3 above)
Comment (davidcok) [reply README.md:73]: Moving .github/CONTRIBUTING to the root folder.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+## Reporting issues
+
+Before reporting an issue here, consider whether it would be better handled in one of the following places:
+
+- [Stack Overflow](https://stackoverflow.com/questions/tagged/dafny), an online Q&A website, for help with writing and proving Dafny programs.
+- [`dafny-lang/ide-vscode`](https://github.com/dafny-lang/ide-vscode), Dafny's VSCode plugin, for issues or feature requests specific to the plugin itself and issues with Dafny's LSP server.
+- [`boogie-org/boogie-friends`](https://github.com/boogie-org/boogie-friends/), Dafny's Emacs mode, for Emacs-specific problems.
+- [`boogie-org/boogie`](https://github.com/boogie-org/boogie), Dafny's intermediate verification language, for questions and issues specific to Boogie.
+
+Other issues can be reported to the [Dafny project](https://githubcom/dafny-lang/dafny/Issues).
```
Comment (robin-aws) [inline CONTRIBUTING.md:15]:
```suggestion
Other issues can be reported to the [Dafny project](https://github.com/dafny-lang/dafny/issues).
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+Thanks for contributing to Dafny!  Github is the right place to discuss feature requests, report issues with Dafny, or contact the Dafny developers.
+Dafny is an Open Source project and welcomes contributions.
+
+## Reporting issues
+
+Before reporting an issue here, consider whether it would be better handled in one of the following places:
+
+- [Stack Overflow](https://stackoverflow.com/questions/tagged/dafny), an online Q&A website, for help with writing and proving Dafny programs.
+- [`dafny-lang/ide-vscode`](https://github.com/dafny-lang/ide-vscode), Dafny's VSCode plugin, for issues or feature requests specific to the plugin itself and issues with Dafny's LSP server.
```
Comment (robin-aws) [inline CONTRIBUTING.md:11]: Don't we still want to manage issues/requests for the LSP server here (as in the old version)?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+## Contributing to the source code
+
+Developers who would like to experiment with and perhaps modify Dafny source should note these resources:
+- First, there are [resources to learn about Dafny as a user](https://dafny.org/dafny/).
+- And note the [community Code of Conduct](https://github.com/dafny-lang/dafny/blob/master/.github/CODE_OF_CONDUCT.md) expected of all participating in the project.
+- The Dafny project is at [github](https://github.com/dafny-lang/dafny) with related [sibling repositories](https://github.com/dafny-lang).
+- Instructions for compiling Dafny are on the [wiki](https://github.com/dafny-lang/dafny/wiki/INSTALL#building-and-developing-from-source-code)
+
+
+If you want to feed code changes for incorporation into the Dafny project, please note:
```
Comment (robin-aws) [inline CONTRIBUTING.md:27]:
```suggestion
If you want to propose code changes for the Dafny project, please note:
```

(just found this phrasing a bit awkward, feel free to ignore)

File: README.md (hunk 2 above)
Comment (robin-aws) [reply README.md:74]: Would still like this

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+Other issues can be reported to the [Dafny project](https://githubcom/dafny-lang/dafny/Issues).
+Please include your Dafny version number, any relevant code, and expected results in the issue report.
+
+## Contributing to the source code
+
+Developers who would like to experiment with and perhaps modify Dafny source should note these resources:
+- First, there are [resources to learn about Dafny as a user](https://dafny.org/dafny/).
+- And note the [community Code of Conduct](https://github.com/dafny-lang/dafny/blob/master/.github/CODE_OF_CONDUCT.md) expected of all participating in the project.
+- The Dafny project is at [github](https://github.com/dafny-lang/dafny) with related [sibling repositories](https://github.com/dafny-lang).
```
Comment (cpitclaudel) [inline CONTRIBUTING.md:23]: Right, but how much does it differ from the exit code returned by the main method?

File: CONTRIBUTING.md (hunk 5 above)
Comment (davidcok) [reply CONTRIBUTING.md:11]: Ah - the original text was ambiguous (at least to me) and I clarified it in the wrong direction.

File: CONTRIBUTING.md (hunk 7 above)
Comment (davidcok) [reply CONTRIBUTING.md:23]: OK...

File: README.md (hunk 2 above)
Comment (davidcok) [reply README.md:74]: Robin, the CONTRIBUTING.md page refers to a wiki page about the process for contributing PRs, which also refers to https://github.com/dafny-lang/dafny/wiki/INSTALL#building-and-developing-from-source-code,which is the one location for all dependencies, including the pre-commit installation (which is only needed if one is going to propose a PR). So do you really want to call out one dependency, the pre-commit script, for special mention?

File: CONTRIBUTING.md (hunk 5 above)
Comment (davidcok) [reply CONTRIBUTING.md:11]: Fixed.

File: README.md (hunk 2 above)
Comment (robin-aws) [reply README.md:74]: Ah I didn't realize it was also documented there. In that case let's just remove this copy of the content about the pre-commit entirely.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,38 +1,53 @@
...
-See [`CODE_OF_CONDUCT.md`](./CODE_OF_CONDUCT.md).
+    Each `.dfy` file in `Test/` is a test, with a  [`lit`](https://llvm.org/docs/CommandGuide/lit.html) header describing how to run it and a `.expect` file indicating the expected output.  See [`Test/README.md`](../Test/README.md) for more info on running Dafny' integration tests.
 
-## Pull requests
+  - Dafny's unit tests are in various `*.Test` directories in [`Source`](../Source).
+
+  Our CI is configured to run all tests when you create a PR.  To run tests locally, use `dotnet test Source/Dafny.sln` (but note that running the tests for our compiler backends requires installing lots of additional software).
 
-Pull requests should be merged using 'Squash and merge' unless otherwise noted. 
+Pull Requests should be merged using 'Squash and merge' unless otherewise noted.
```
Comment (robin-aws) [inline CONTRIBUTING.md:47]:
```suggestion
Pull Requests should be merged using 'Squash and merge' unless otherwise noted.
```

**Review by robin-aws — APPROVED** [review]

**Pull request merged**